### PR TITLE
Processor resource discovery

### PR DESCRIPTION
**Pull request by kba**
Implementing https://github.com/OCR-D/spec/pull/163

* processors have a new method `resolve_resource` that gets passed a parameter value and tries to find it in one of the locations defined in OCR-D/spec#163
* core now has a builtin list (extensible via user list) of known resources for processors, e.g. the gt4histocr models for tesseract and calamari
* new CLI `ocrd resmgr` that allows browsing the builtin list, listing installed resources and downloading new resources
* resources that are listed in the builtin list can be referenced by URL or name

### CODE REVIEW

File: ocrd/ocrd/resource_manager.py
Comment (bertsky) [inline ocrd/ocrd/resource_manager.py:64]: I'm confused: What exact circumstances would lead us into this branch? 

My test already fails earlier: After…

    ocrd resmgr download -n ocrd-cis-ocropy-recognize https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz

…I still get _no_ results for…

    ocrd resmgr list-installed

…(because the unregistered download does not touch the user resources – see other comment – and the outer loop only looks for registered ones), but…

    ocrd resmgr list-installed -e ocrd-cis-ocropy-recognize

…already fails with…

```
  File "venv/lib/python3.6/site-packages/ocrd/cli/resmgr.py", line 50, in list_installed
    for executable, reslist in resmgr.list_installed(executable):
  File "venv/lib/python3.6/site-packages/ocrd/resource_manager.py", line 62, in list_installed
    resdict = [x for x in self.database[executable] if x['name'] == res_name]
KeyError: 'ocrd-cis-ocropy-recognize'
```

So it would seem that you have to enter either via a loaded and validated resource database (in which case IIUC there will always be an entry for any of its resources), or just by name (in which case IIUC no database lookup is warranted anyway)...

File: ocrd/ocrd/processor/base.py
Comment (bertsky) [inline ocrd/ocrd/processor/base.py]: Why did you remove the `pkg_resources.resource_filename` (Python fall-back) mechanism entirely? Isn't this even required by the spec now?

(I know it's harder to get these under the centralized `ocrd resmgr list-installed`. But if `Processor.list_all_resources()` would append the files from pkg_resources, then the former could delegate to these via CLI calls.)

File: ocrd/ocrd/processor/base.py
Comment (bertsky) [inline ocrd/ocrd/processor/base.py:183]: Plus append by anything in `pkg_resources.resource_filename` (see comment on `resolve_resource`...)?

File: ocrd_models/ocrd_page_user_methods.py
Comment (bertsky) [inline ocrd_models/ocrd_page_user_methods.py:116]: By analogy to our `get_AllRegions`, IMO this should also be plural here:

```suggestion
    _add_method(r'^(PageType)$', 'get_AllTextLines'),
```

File: ocrd_models/ocrd_page_user_methods/get_AllTextLine.py
Comment (bertsky) [inline ocrd_models/ocrd_page_user_methods/get_AllTextLine.py:9]: Besides the plural, I recommend also offering an `order` kwarg here, delegating to `get_AllRegions` at the region level, and additionally respecting `textLineOrder` at the line level.

**Review by bertsky — COMMENTED** [review]
Very impressive!

Not sure I understand all ramifications yet. What would need to be adapted
1. processors code
2. model deployment in ocrd_all and modules
3. documentation (workflows, usage, installation)

For 1 and 2, it would help to see an example, say for ocrd_tesserocr.

For 2, there's also the question of how to incorporate Python package resources and git submodules.

Your beta pre-release is an excellent vehicle for realistic tests in that direction.

But let's first sort out some details (in the comments)...

**Comment by kba** [comment]
> Not sure I understand all ramifications yet. What would need to be adapted
> 
> 1. processors code

Basically, replace `self.parameter['my-file-parameter']` with `self.resolve_resource(self.parameter['my-file-parameter'])`. 

> 2. model deployment in ocrd_all and modules

`make install-models*` should be adapted to store models in the search path (e.g. use `venv/share/ocrd-calamari-recognize` instead of `venv/share/ocrd_calamari/`. 

> 3. documentation (workflows, usage, installation)

The models documentation would need a rewrite with examples of `ocrd resmgr` and an explanation of the search path logic. Also, we should decide on a convention for docker, e.g. mounting a host directory of models as `$HOME/.cache`.

> 
> For 1 and 2, it would help to see an example, say for ocrd_tesserocr.
> 

Here's the change for ocrd_calamari: https://github.com/OCR-D/ocrd_calamari/pull/53/files, here for ocrd_anybaseocr https://github.com/OCR-D/ocrd_anybaseocr/pull/78/files


> For 2, there's also the question of how to incorporate Python package resources and git submodules.

I'm not sure I understand - can you elaborate pls?

File: ocrd/ocrd/resource_manager.py
Comment (kba) [reply ocrd/ocrd/resource_manager.py:18]: Agreed, `$HOME/ocrd/resources.yml` was a typo anyway.

File: ocrd/ocrd/cli/resmgr.py
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: > IIUC this means the new / unregistered resources will not be visible under list-installed afterwards.

They will, but currently there's just question marks where the description and URL would be for a registered resource:

```
$ touch $HOME/.cache/ocrd-tesserocr-recognize
$ ocrd resmgr list-installed
ocrd-tesserocr-recognize
- bar (???)
  ???
- Fraktur_GT4HistOCR.traineddata (https://ub-backup.bib.uni-mannheim.de/~stweil/ocrd-train/data/Fraktur_5000000/tessdata_fast/Fraktur_50000000.334_450937.traineddata)                       
  Tesseract LSTM model trained on GT4HistOCR
```

You mean, create an entry with URL and name in `$HOME/.config/resources.yml` and create that if it doesn't exist yet? Yes that would work.

> ```ocrd resmgr download -n ocrd-cis-ocropy-recognize https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz```

My reasoning was: If you use a URL once, you will use this URL in the future (and not the short name), hence the mapping from URL to name should be simple to make this reproducible. But I guess "last URL segment sans fragment identifier and query" would work just as well.

File: ocrd/ocrd/cli/resmgr.py
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:82]: > For example, I might have mounted my models via NFS or downloaded them via some other means already

You mean, you want to allow absolute file paths in addition to HTTP(S) URL? Sure, that would be doable.

**Comment by bertsky** [comment]
> > For 2, there's also the question of how to incorporate Python package resources and git submodules.
> 
> I'm not sure I understand - can you elaborate pls?

Modules might want to distribute config files or models via Python package resources. This should be standardized for OCR-D and supported by the resmgr / Processor reference implementation.

The git submodules part refers to how can we generalize ocrd_all's `install-models` to conveniently cover other modules, with models either under some makefile target or in a data submodule.

File: ocrd/ocrd/resource_manager.py
Comment (kba) [reply ocrd/ocrd/resource_manager.py:64]: > …I still get no results for…

The problem is: what do we iterate? Do we iterate over all the items in all the XDG_FOLDERS that follow our convention (`ocrd-.*`) or over the keys in the `resource_list.yml`? I chose the latter for simplicity's sake but I understand how that can be confusing.

The branch is triggered if there is a resource for one of the processors for which we know there are resources (has a key in `resource_list.yml`) but not the resource with this particular name.

I'll change the outer loop to actually inspect the file system.

File: ocrd/ocrd/cli/resmgr.py
Comment (bertsky) [reply ocrd/ocrd/cli/resmgr.py:89]: > You mean, create an entry with URL and name in `$HOME/.config/resources.yml` and create that if it doesn't exist yet? Yes that would work.

Yes, exactly that.

> > `ocrd resmgr download -n ocrd-cis-ocropy-recognize https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz`
> 
> My reasoning was: If you use a URL once, you will use this URL in the future (and not the short name), hence the mapping from URL to name should be simple to make this reproducible. But I guess "last URL segment sans fragment identifier and query" would work just as well.

You mean the user will pass the URL in the parameter JSON? But then why does it say 
> Use in parameters as 'httpsgithubcomzuphilipocropymodelsrawmasterendefaultpyrnngz'

? I would expect either 
> Use in parameters as 'https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz'

or (better)
> Use in parameters as 'en-default.pyrnn.gz'

File: ocrd/ocrd/processor/base.py
Comment (kba) [reply ocrd/ocrd/processor/base.py]: > Why did you remove the pkg_resources.resource_filename (Python fall-back) mechanism again entirely? Isn't this even required by the spec now?

To be honest, I was trying to get this right for ocrd_typegroups_classifier but I found it more intuitive to offer the bundled resource as a regular resource in `resources_list.yml` (since it is already in GitHub and hence HTTP-downloadable) and let ocrd_typegroups_classifier handle the case where no nerwork parameter was given itself. It's a bit of a waste of space but makes the code for `list_all_resources` and `resmgr` simpler.

It would be good to have an example project that makes use of bundled resources to test this with and if we don't, it would be better to change the spec TBH. 

BTW, I wanted to also list https://github.com/ASVLeipzig/cor-asv-ann-models but gave up on getting a "lightweight" solution working to download from LFS.

File: ocrd_models/ocrd_page_user_methods.py
Comment (kba) [reply ocrd_models/ocrd_page_user_methods.py:116]: yep, it's also `get_AllAlternativeImages` now, I'll fix this.

File: ocrd/ocrd/cli/resmgr.py
Comment (bertsky) [reply ocrd/ocrd/cli/resmgr.py:82]: Yes, absolute or CWD-relative paths – it does not matter, because you would have to copy (or link) them to a standard location (where `list_all_resources` would expect them) anyway.

File: ocrd/ocrd/cli/resmgr.py
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: > `Use in parameters as 'httpsgithubcomzuphilipocropymodelsrawmasterendefaultpyrnngz'`

Excellent point, my reasoning was flawed :)

File: ocrd/ocrd/processor/base.py
Comment (bertsky) [reply ocrd/ocrd/processor/base.py]: > > Why did you remove the pkg_resources.resource_filename (Python fall-back) mechanism again entirely? Isn't this even required by the spec now?
> 
> To be honest, I was trying to get this right for ocrd_typegroups_classifier but I found it more intuitive to offer the bundled resource as a regular resource in `resources_list.yml` (since it is already in GitHub and hence HTTP-downloadable) and let ocrd_typegroups_classifier handle the case where no nerwork parameter was given itself. It's a bit of a waste of space but makes the code for `list_all_resources` and `resmgr` simpler.

I see. Well, `ocrd_typegroups_classifier` or [document-image-binarization](https://github.com/ajgallego/document-image-binarization) are examples of packages which distribute their model files (not sure about binary vs sdist installation though), so it _should_ already be available locally. But I agree that the waste of bandwidth/space is probably not a big deal in general because if the models are too large, devs will not include them in the package (or even the same repo) anyway.

For modules with very large files like `ocrd_anybaseocr` – not all of which might be needed – I guess it makes no sense to expect them to be distributed along with the base pkg. (Although the PyPI server could still help distribute dependent packages like `ocrd-anybaseocr-models-dewarping`, `ocrd-anybaseocr-models-segmentation` etc.)

But then there's modules like ocrd_wrap: I want to put many small preset files there in such a way that they can be referenced on the CLI  (`-p param_im6convert-denoise-raw.json -P level-of-operation region`). But if that's (going to be) possible by just doing `ocrd resmgr download ocrd-preprocess-image '*'` (which I could even delegate to in my `make install` rule), then fine. (I might even register resources for external tools there.)

> It would be good to have an example project that makes use of bundled resources to test this with and if we don't, it would be better to change the spec TBH.

Yes, agreed: If we don't actually need package resources as a mechanism, we shouldn't have it in the spec. Wrapping an existing pkg resource to an OCR-D file resource spec should be easy.

But that's actually something I still don't see here: we have a _centralized_ `resource_list.yml` now, but not a way for modules to feed their contributions...

> BTW, I wanted to also list https://github.com/ASVLeipzig/cor-asv-ann-models but gave up on getting a "lightweight" solution working to download from LFS.

I'll have a look myself later-on. Github LFS always gave us problems in the past already (we usually exceed the monthly quota without even knowing why). There's still the Gitlab backup, but...

File: ocrd_models/ocrd_page_user_methods/get_AllTextLine.py
Comment (kba) [reply ocrd_models/ocrd_page_user_methods/get_AllTextLine.py]: 4def1d97, implemented a `region_order` kwarg delegating to `order` in `get_AllRegions`. As for the `textLineOrder`, I've asked @tboenig to clarify what this means and how we can test it.

File: ocrd_models/ocrd_page_user_methods/get_AllTextLines.py
Comment (bertsky) [inline ocrd_models/ocrd_page_user_methods/get_AllTextLines.py:14]: I think the mnemonics should be different, `textline_order` should be a boolean.

If true, then do something like:
```python
    for reg in self.get_AllRegions(['Text'], order=region_order):
        lo = reg.get_textLineOrder() or self.get_textLineOrder() or 'top-to-bottom'
        lines = reg.get_TextLine()
        ret += lines if lo in ['top-to-bottom', 'left-to-right'] else list(reversed(lines))
    return ret
```

Not entirely sure about the PAGE-XML semantics here TBH. Must XML ordering always be top-to-bottom (or left-to-right in case of vertical text)? In that case, you would need to _apply_ `@textLineOrder` to get the correct text concatenation (`right-to-left` for classical Chinese). But the attribute could also be interpreted to _describe_ just how the current XML ordering looks like (which you could verify by its coordinates). In that case, you would never re-order anything.

File: ocrd_models/ocrd_page_user_methods/get_AllTextLine.py
Comment (bertsky) [reply ocrd_models/ocrd_page_user_methods/get_AllTextLine.py]: > As for the `textLineOrder`, I've asked @tboenig to clarify what this means and how we can test it.

Ah, great – I was just going to ask you for that, cf. [this comment](https://github.com/OCR-D/core/pull/559/files/4bf12fb7eaa39fc2a73874df3b69c4b6b24b5965..4def1d974e9645753db924b5ca0c5ed1e7682ed5#r551472862)

File: ocrd/ocrd/resource_manager.py
Comment (kba) [reply ocrd/ocrd/resource_manager.py:76]: 79116038

File: ocrd/ocrd/cli/resmgr.py
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: BTW, https://github.com/zuphilip/ocropy-models/raw/master/en-default.pyrnn.gz is already in our resource_list.yml, I just had a typo in it, fixed in 54a214a0

File: ocrd/ocrd/resource_manager.py
Comment (bertsky) [inline ocrd/ocrd/resource_manager.py:69]: This does cover all circumstances for unregistered resources I could think of, but it also includes false positives now:
```
ocrd-im6convert
- ocrd-tool.json (???)
  ???
```

It seems that `venv/share/ocrd-im6convert/ocrd-tool.json` matched the `startswith` pattern here. So require `os.path.isdir(x)`, too?

Comment (kba) [reply ocrd/ocrd/resource_manager.py:69]: Yeah, I saw that too. But `isdir(x)` wouldn't help because `ocrd-im6convert` is a directory. I think I'll just make a new im6convert release with `SHAREDIR=$VIRTUAL_ENV/share/ocrd_im6convert` (note the `_` instead of `-`).

File: ocrd/ocrd/resource_manager.py
Comment (kba) [reply ocrd/ocrd/resource_manager.py]: e33346b4 New behavior: Look up resources with the builtin list and user list at `XDG_CONFIG_HOME/ocrd/resources.yml`. If an unknown resource is encountered for the first time, creates a stub entry in the user list.

File: ocrd/ocrd/resource_manager.py
Comment (bertsky) [reply ocrd/ocrd/resource_manager.py:76]: Yes, that makes sense. (ocrd_olena, ocrd_pagetopdf and ocrd_fileformat all use the underscore already.)

File: ocrd/ocrd/cli/resmgr.py
Comment (bertsky) [reply ocrd/ocrd/cli/resmgr.py:89]: Maybe (besides `nth_url_segment`) even normalize URL encoding (eg. replace `%C5%BF` by `ſ`)?

Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: 3ee66ce5 Now:

```
ocrd resmgr download -n ocrd-foo https://github.com/zuphilip/ocropy-french-models/raw/master/models/french-balzac-model-00105000.pyrnn.gz                                                 
15:13:11.315 INFO ocrd.cli.resmgr - No resources found in registry
15:13:11.315 INFO ocrd.cli.resmgr - Downloading unregistered resource https://github.com/zuphilip/ocropy-french-models/raw/master/models/french-balzac-model-00105000.pyrnn.gz               
  [####################################]  100%
15:13:16.579 INFO ocrd.cli.resmgr - Downloaded https://github.com/zuphilip/ocropy-french-models/raw/master/models/french-balzac-model-00105000.pyrnn.gz to /home/kba/.cache/ocrd-foo/french-ba
lzac-model-00105000.pyrnn.gz
15:13:16.579 INFO ocrd.cli.resmgr - Use in parameters as 'french-balzac-model-00105000.pyrnn.gz'

```

Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: > Maybe (besides nth_url_segment) even normalize URL encoding (eg. replace %C5%BF by ſ), perhaps even dropping all query/fragment components from the path?

Good point b21d4628 Do you have an example URL to test with?

Comment (bertsky) [reply ocrd/ocrd/cli/resmgr.py:89]: I do – and it _fails_ because the HEAD request does not get answered with a `content-length` header in Gitlab:
- my [s-ſ reconstruction model](https://git.informatik.uni-leipzig.de/ocr-d/cor-asv-ann-models/-/raw/master/s2s.gt4histocr.s-%C5%BF.d2.w0512.adam.attention.stateless.variational-dropout.char.transfer-lm.h5?inline=false)

```
  File "venv/lib/python3.6/site-packages/ocrd/cli/resmgr.py", line 85, in download
    content_length = int(r.headers.get('content-length'))
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

See [here](https://stackoverflow.com/a/44299915/14474237) for a solution (i.e. use GET but with `stream=True`)...

Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: Thanks, I will test with that and replace HEAD with bodyless GET. And good to know these resources are mirrorred on your GitLab, I was trying to get support for Git LFS but it is surprisingly tricky to implement without additional dependencies or lots of copy/paste code.

Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:89]: d8d97af1 

```
13:38:08.997 INFO ocrd.cli.resmgr - Downloaded https://git.informatik.uni-leipzig.de/ocr-d/cor-asv-ann-models/-/raw/master/s2s.gt4histocr.s-%C5%BF.d2.w0512.adam.attention.stateless.variation
al-dropout.char.transfer-lm.h5?inline=false to /home/kba/.cache/ocrd-foo/s2s.gt4histocr.s-ſ.d2.w0512.adam.attention.stateless.variational-dropout.char.transfer-lm.h5                         
13:38:08.998 INFO ocrd.cli.resmgr - Use in parameters as 's2s.gt4histocr.s-ſ.d2.w0512.adam.attention.stateless.variational-dropout.char.transfer-lm.h5'
```

Comment (bertsky) [reply ocrd/ocrd/cli/resmgr.py:89]: Fantastic!

File: ocrd_models/ocrd_page_user_methods/get_AllTextLine.py
Comment (bertsky) [reply ocrd_models/ocrd_page_user_methods/get_AllTextLine.py]: cf. https://github.com/PRImA-Research-Lab/PAGE-XML/issues/26

File: ocrd/ocrd/cli/resmgr.py
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py]: 509200c8

File: ocrd/ocrd/processor/base.py
Comment (kba) [reply ocrd/ocrd/processor/base.py]: > But that's actually something I still don't see here: we have a centralized resource_list.yml now, but not a way for modules to feed their contributions...

There is the option of a user-defined list in addition to the centralized list. Another option would be to extend the ocrd-tool.json schema so that project authors can themselve define the resources and we introspect that at runtime?

Comment (bertsky) [reply ocrd/ocrd/processor/base.py]: > Another option would be to extend the ocrd-tool.json schema so that project authors can themselve define the resources and we introspect that at runtime?

Excellent idea! So ocrd_all installation rules could aggregate ocrd's `resource_list.yml` or `~/.config/ocrd/resource.yml` from that (enabling `make install-models` after `make all`). But also, each individual module makefile could get a local `install-models` rule...

File: ocrd/ocrd/processor/base.py
Comment (bertsky) [reply ocrd/ocrd/processor/base.py:182]: > Plus append by anything in `pkg_resources.resource_filename` (see comment on `resolve_resource`...)?

We don't want that anymore (https://github.com/OCR-D/spec/commit/fa6e00a269713d8593e77f9b2a3d8f19560e7649)

File: ocrd/ocrd/processor/base.py
Comment (kba) [reply ocrd/ocrd/processor/base.py]: Since integrating this into ocrd-tool.json requires changes to the spec and we should probably adapt the metadata scheme to be interoperable with [@mittagessen's schema](https://github.com/mittagessen/kraken/blob/master/kraken/metadata.schema.json) I will do that in a separate set of PRs.

File: ocrd_models/ocrd_page_user_methods/get_AllTextLines.py
Comment (kba) [reply ocrd_models/ocrd_page_user_methods/get_AllTextLines.py:14]: Thanks, I'll implement it like that for now until we agreed on the right interpretation in https://github.com/PRImA-Research-Lab/PAGE-XML/issues/26

File: ocrd/ocrd/processor/base.py
Comment (bertsky) [reply ocrd/ocrd/processor/base.py]: So IIUC you want to merge the centralized resource list first, and adapt ocrd_all, and then work towards decentralized, aggregated resource definitions in tool.json?

But the Kraken model schema does not even have a URL attribute! Or are you planning for some schema extension? And isn't it specific to OCR models (`script`, `graphemes`)? Also, `accuracy` by itself does not make sense to me. At least, some of these features should not be `required`...

Comment (bertsky) [reply ocrd/ocrd/processor/base.py]: accuracy: Yes, the method / metric often is a parameter itself, but more importantly, different models will usually also have different test data...

Kraken, Okralact, Calamari config, ... so many options. I don't think we should even try to use some interoperable / unified scheme.

File: ocrd_models/ocrd_page_user_methods/get_AllTextLines.py
Comment (kba) [reply ocrd_models/ocrd_page_user_methods/get_AllTextLines.py]: 012e49ea The kwarg is now `respect_textline_order` and the algorithm like you proposed.

**Review by bertsky — COMMENTED** [review]
If I could make another wish: It would be useful if `list-installed` would show the location type registered ressources were found in (`data`, `config`, `cache`, `virtualenv`), and likewise for unregistered resources (instead of the full path after `Found at`).

**Comment by kba** [comment]
> If I could make another wish: It would be useful if list-installed would show the location type registered ressources were found in (data, config, cache, virtualenv), and likewise for unregistered resources (instead of the full path after Found at).

Sure, that makes sense.

**Comment by kba** [comment]
> If I could make another wish: It would be useful if `list-installed` would show the location type registered ressources were found in (`data`, `config`, `cache`, `virtualenv`), and likewise for unregistered resources (instead of the full path after `Found at`).

a3cff9e8

```
$ ocrd resmgr download -n ocrd-tesserocr-recognize https://github.com/tesseract-ocr/tessdata/raw/master/fin.traineddata

$ cat $HOME/.config/ocrd/resources.yml
cat ~/.config/ocrd/resources.yml 
# OCR-D private resource list (consider sending a PR with your own resources to OCR-D/core)
ocrd-tesserocr-recognize:
- description: Found at virtualenv on 2021-01-21 19:18:30.631387
  name: fin.traineddata
  size: 21140513
  url: https://github.com/tesseract-ocr/tessdata/raw/master/fin.traineddata
  version_range: ???

$ ocrd resmgr list-installed -e ocrd-tesserocr-recognize                                                               
ocrd-tesserocr-recognize
- fin.traineddata @ virtualenv (https://github.com/tesseract-ocr/tessdata/raw/master/fin.traineddata)
  Found at virtualenv on 2021-01-21 19:18:30.63138
```

**Comment by kba** [comment]
OK, so adapted to what @bersky and I discussed and up-to-date with https://github.com/OCR-D/spec/pull/178 🤞

File: ocrd/ocrd/resource_manager.py
Comment (bertsky) [inline ocrd/ocrd/resource_manager.py:73]: What about the CWD? I know it's probably not realistic, because at processor runtime we are in the workspace, whereas during `ocrd resmgr` we might not, but we could still try :-)

```suggestion
            parent_dirs = [join(x, 'ocrd-resources') for x in [XDG_DATA_HOME, '/usr/local/share', getcwd()]]
```

**Review by bertsky — COMMENTED** [review]
Maybe we could also import and expose `OcrdResourceManager` directly from `ocrd`, so use-cases like ocrd_tesserocr are more accessible.

File: ocrd/ocrd/resource_manager.py
Comment (bertsky) [inline ocrd/ocrd/resource_manager.py:133]: Also, it might still be helpful to encapsulate the default path here (think ocrd_tesserocr again):

```suggestion

    @property
    def default_resource_dir(self):
        return self.location_to_resource_dir('data')
```

**Review by bertsky — CHANGES_REQUESTED** [review]

File: ocrd/ocrd/resource_manager.py
Comment (kba) [reply ocrd/ocrd/resource_manager.py:73]: Originally omitted because we didn't have the `ocrd-resources` directory and listing everything in PWD seemed overkill. But there's no reason now not include it.

**Comment by kba** [comment]
> Maybe we could also import and expose OcrdResourceManager directly from ocrd, so use-cases like ocrd_tesserocr are more accessible.

a5858ec4

I also added a flag `--allow-uninstalled` to control whether to allow downloading resources for processors that are not installed. Without it `ocrd resmgr ocrd-foo /vmlinuz` will now cause an error (unless you have `ocrd-foo` in your `$PATH`).

Also added `ocrd resmgr download '*`` which will download all resources for all installed processors (or all resources in the registry if `--allow-uninstalled`).

I really hope we've got it right this time but I will test some more before merging just to be sure.

File: ocrd/ocrd/cli/resmgr.py
Comment (kba) [reply ocrd/ocrd/cli/resmgr.py:5]: > but it also means to rule out processors/resources as installed if they are behind a sub-venv again. 

How do you mean? If the wrapper scripts are in `$PATH`, then why can we not assume that they are installed?

Comment (bertsky) [reply ocrd/ocrd/cli/resmgr.py:5]: I'm sorry – you can scratch that entire comment. (I forgot this already _is_ search in `PATH`, not in Python packages. Also, I already used that exact pattern in the workflow server...)

**Review by bertsky — APPROVED** [review]

**Pull request merged**
